### PR TITLE
Bug 1986003: Retry kubeconfig checks, when kube-apiserver is temporarily unavailable

### PR DESCRIPTION
**Pull request by soltysh**
/assign @p0lyn0mial 

since you complained about this particular to me some time ago, maybe not necessarily about this particular problem but it's a good starting point :wink: which I noticed fails pretty frequently

### CODE REVIEW

File: test/extended/apiserver/kubeconfigs.go [hunk 1]
```diff
@@ -29,13 +32,25 @@ var _ = g.Describe("[Conformance][sig-api-machinery][Feature:APIServer] local ku
...
+
+func testNode(oc *exutil.CLI, kubeconfig, masterName string) (bool, error) {
+	g.By("Testing master node " + masterName)
+	kubeconfigPath := "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/" + kubeconfig
+	framework.Logf("Verifying kubeconfig %q on master %q", kubeconfig, masterName)
+	out, err := oc.AsAdmin().Run("debug").Args("node/"+masterName, "--", "chroot", "/host", "/bin/bash", "-euxo", "pipefail", "-c",
+		fmt.Sprintf(`oc --kubeconfig "%s" get namespace kube-system`, kubeconfigPath)).Output()
+	framework.Logf(out)
+	// retry error when kube-apiserver was temporarily unavailable
+	retry := strings.Contains(out, "The connection to the server localhost:6443 was refused")
```
Comment (p0lyn0mial) [inline test/extended/apiserver/kubeconfigs.go:54]: would you be open to adding a precondition before running the test instead? 
not sure this is the only error we might get, for example on an IPv6 cluster we might get `[::1]:6443: connect: connection refused`

we could check if the cluster is in stable condition (not progressing, the pods on the same revision) for X min.

**Comment by p0lyn0mial** [comment]
> since you complained about this particular to me some time ago, maybe not necessarily about this particular problem but it's a good starting point 😉 which I noticed fails pretty frequently

wasn't me :)

File: test/extended/apiserver/kubeconfigs.go [hunk 2]
```diff
@@ -29,13 +32,25 @@ var _ = g.Describe("[Conformance][sig-api-machinery][Feature:APIServer] local ku
 			framework.Logf("Discovered %d master nodes.", len(masterNodes.Items))
 			o.Expect(masterNodes.Items).NotTo(o.HaveLen(0))
 			for _, master := range masterNodes.Items {
-				g.By("Testing master node " + master.Name)
-				kubeconfigPath := "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/" + kubeconfig
-				framework.Logf("Verifying kubeconfig %q on master %s", master.Name)
-				out, err := oc.AsAdmin().Run("debug").Args("node/"+master.Name, "--", "chroot", "/host", "/bin/bash", "-euxo", "pipefail", "-c", fmt.Sprintf(`oc --kubeconfig "%s" get namespace kube-system`, kubeconfigPath)).Output()
+				retry, err := testNode(oc, kubeconfig, master.Name)
+				if retry {
```
Comment (aojea) [inline test/extended/apiserver/kubeconfigs.go:36]: is one retry enough?

Comment (soltysh) [reply test/extended/apiserver/kubeconfigs.go:36]: yeah, after consideration I did 2 :wink:

**Comment by soltysh** [comment]
@aojea updated, ptal

File: test/extended/apiserver/kubeconfigs.go (hunk 1 above)
Comment (soltysh) [reply test/extended/apiserver/kubeconfigs.go:57]: Checking progressing might slow this down, but we can add that as last resort, re-tries seems simpler, b/c theoretically we can even pass the test when rollout is in progress.

Comment (soltysh) [reply test/extended/apiserver/kubeconfigs.go:57]: Added check for ipv6, though

**Comment by soltysh** [comment]
or @p0lyn0mial

File: test/extended/apiserver/kubeconfigs.go [hunk 3]
```diff
@@ -29,13 +32,29 @@ var _ = g.Describe("[Conformance][sig-api-machinery][Feature:APIServer] local ku
...
+func testNode(oc *exutil.CLI, kubeconfig, masterName string) (bool, error) {
+	g.By("Testing master node " + masterName)
+	kubeconfigPath := "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/" + kubeconfig
+	framework.Logf("Verifying kubeconfig %q on master %q", kubeconfig, masterName)
+	out, err := oc.AsAdmin().Run("debug").Args("node/"+masterName, "--", "chroot", "/host", "/bin/bash", "-euxo", "pipefail", "-c",
+		fmt.Sprintf(`oc --kubeconfig "%s" get namespace kube-system`, kubeconfigPath)).Output()
+	framework.Logf(out)
+	// retry error when kube-apiserver was temporarily unavailable
+	retry := strings.Contains(out, "The connection to the server localhost:6443 was refused") ||
+		strings.Contains(out, "[::1]:6443: connect: connection refused")
```
Comment (aojea) [inline test/extended/apiserver/kubeconfigs.go:58]: what are the messages so different between protocols? is just curiosity , using curl they are more consisten
```
$ curl localhost:6443
curl: (7) Failed to connect to localhost port 6443: Connection refused
$ curl [::1]:6443
curl: (7) Failed to connect to ::1 port 6443: Connection refused
$ curl 127.0.0.1:6443
curl: (7) Failed to connect to 127.0.0.1 port 6443: Connection refused
```

**Comment by soltysh** [comment]
> lgtm, just a question to double check the "retry error messages"

I just checked that, you're right the message should be identical, only the address will change, it's coming from here: https://github.com/kubernetes/kubernetes/blob/cbb5ea8210596ada1efce7e7a271ca4217ae598e/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L237-L243, so I've updated accordingly the PR.  

With:
```golang

File: test/extended/apiserver/kubeconfigs.go [hunk 4]
```diff
@@ -29,13 +32,29 @@ var _ = g.Describe("[Conformance][sig-api-machinery][Feature:APIServer] local ku
...
+
+func testNode(oc *exutil.CLI, kubeconfig, masterName string) (bool, error) {
+	g.By("Testing master node " + masterName)
+	kubeconfigPath := "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/" + kubeconfig
+	framework.Logf("Verifying kubeconfig %q on master %q", kubeconfig, masterName)
+	out, err := oc.AsAdmin().Run("debug").Args("node/"+masterName, "--", "chroot", "/host", "/bin/bash", "-euxo", "pipefail", "-c",
+		fmt.Sprintf(`oc --kubeconfig "%s" get namespace kube-system`, kubeconfigPath)).Output()
+	framework.Logf(out)
+	// retry error when kube-apiserver was temporarily unavailable, this matches oc error coming from:
+	// https://github.com/kubernetes/kubernetes/blob/cbb5ea8210596ada1efce7e7a271ca4217ae598e/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L237-L243
```
Comment (aojea) [inline test/extended/apiserver/kubeconfigs.go:57]: the message you've linked ends with a quesiton tag, but the regex work fine
https://play.golang.org/p/aBGVR3_3YsQ

Comment (soltysh) [reply test/extended/apiserver/kubeconfigs.go:57]: Yeah, I didn't want to be that precise :wink:

Comment (aojea) [reply test/extended/apiserver/kubeconfigs.go:57]: those symbols on regexps :grimacing: 
:rofl:

**Comment by soltysh** [comment]
/override ci/prow/e2e-agnostic-cmd
/override ci/prow/e2e-aws-single-node
These are not required and I'll continue debugging test-cmd separately.

**Pull request merged**